src/muz/spacer/spacer_context.cpp

```diff
@@ -3320,7 +3332,82 @@ void context::predecessor_eh()
             m_callbacks[i]->predecessor_eh();
     }
 }
+bool pred_transformer::check_models_reach(const datalog::rule* r, model* model) {
+  expr_ref rf(m);
+  reach_fact_ref_vector child_reach_facts;
+
+  SASSERT(r != nullptr);
+  ptr_vector<func_decl> preds;
+  find_predecessors(*r, preds);
+  for (unsigned i = 0; i < preds.size(); i++) {
+    func_decl *pred = preds[i];
+    bool atleast_one_true = false;
+    pred_transformer &ch_pt = ctx.get_pred_transformer(pred);
+    // get all reach facts of tail preds used in the model
+    expr_ref o_ch_reach(m);
+    reach_fact_ref_vector used_rfs;
+    ch_pt.get_all_used_rf(*model, i, used_rfs);
+    for (auto* rf: used_rfs) {
+      pm.formula_n2o(rf->get(), o_ch_reach, i);
+      if (!model->is_true(o_ch_reach)) {
+          func_decl *tag = to_app(rf->tag())->get_decl();
+          model->unregister_decl(tag);
+          model->register_decl(tag, m.mk_true());
+          model->reset_eval_cache();
+      }
+      else atleast_one_true = true;
+    }
+    if(used_rfs.size() > 0 && !atleast_one_true) {
+      TRACE("spacer_detail",
+            tout << "model does not satisfy any reachable fact\n";);
+      return false;
+    }
+  }
+  return true;
+}
+
+bool context::check_models_reach(model* mdl) {
+    reach_fact_ref_vector used_rfs;
+    expr_ref exp(m);
+    for(auto & rel : m_rels) {
+        bool atleast_one_true = false;
+        pred_transformer& pt = *rel.m_value;
+        used_rfs.reset();
+        manager &pm = pt.get_manager();
+        pt.get_all_used_rf(*mdl, pm.n_index(), used_rfs);
```

Comment:
I do not know what this line is doing. There are no reach facts, and therefore, no reach tags, in the model over next-state variables. 

`get_all_used_rf` takes an `oidx` which is an origin index, but it is given an `n_index` which is not. Unfortunately, the compiler does not check this since all indexes are represented as raw integers.